jxl/src/frame/modular/predict.rs

```diff
@@ -152,17 +155,244 @@ impl Predictor {
+        };
+        let te_n = self.error[pos_n];
+        let te_nw = self.error[pos_nw];
+        let sum_wn = te_n + te_w;
+        let te_ne = self.error[pos_ne];
+
+        let abs_te_w = te_w.abs();
+        let p = if te_n.abs() > abs_te_w {
+            te_n
+        } else if te_nw.abs() > abs_te_w {
+            te_nw
+        } else if te_ne.abs() > abs_te_w {
+            te_ne
+        } else {
+            te_w
+        } as i32;
+
+        self.prediction[0] = w + ne - n;
+        self.prediction[1] = n - (((sum_wn + te_ne) * self.header.wp_header.p1c as i64) >> 5);
+        self.prediction[2] = w - (((sum_wn + te_nw) * self.header.wp_header.p2c as i64) >> 5);
+        self.prediction[3] = n
+            - ((te_nw
+                .wrapping_mul(self.header.wp_header.p3ca as i64)
+                .wrapping_add(te_n.wrapping_mul(self.header.wp_header.p3cb as i64))
+                .wrapping_add(te_ne.wrapping_mul(self.header.wp_header.p3cc as i64))
+                .wrapping_add((nn - n).wrapping_mul(self.header.wp_header.p3cd as i64))
+                .wrapping_add((nw - w).wrapping_mul(self.header.wp_header.p3ce as i64)))
+                >> 5);
+
+        self.pred = weighted_average(&self.prediction, &mut weights);
+
+        if ((te_n ^ te_w) | (te_n ^ te_nw)) <= 0 {
+            let mx = w.max(ne.max(n));
+            let mn = w.min(ne.min(n));
+            self.pred = mn.max(mx.min(self.pred));
+        }
+        ((self.pred + PREDICTION_ROUND) >> PRED_EXTRA_BITS, p)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::headers::modular::{GroupHeader, WeightedHeader};
+
+    use super::{PredictionData, WeightedPredictorState};
+
+    struct SimpleRandom {
+        out: i64,
+    }
+
+    impl SimpleRandom {
+        fn new() -> SimpleRandom {
+            SimpleRandom { out: 1 }
+        }
+        fn next(&mut self) -> i64 {
+            self.out = self.out * 48271 % 0x7fffffff;
+            self.out
+        }
+    }
+
+    fn step(
+        rng: &mut SimpleRandom,
+        state: &mut WeightedPredictorState,
+        xsize: usize,
+        ysize: usize,
+    ) -> (i64, i32) {
+        let pos = (rng.next() as usize % xsize, rng.next() as usize % ysize);
+        let res = state.predict_and_property(
+            pos,
+            xsize,
+            &PredictionData {
+                top: rng.next() as i32,
+                left: rng.next() as i32,
+                topright: rng.next() as i32,
+                topleft: rng.next() as i32,
+                toptop: rng.next() as i32,
+                leftleft: 0,
+                toprightright: 0,
+            },
+        );
+        state.update_errors(rng.next(), pos, xsize);
+        res
+    }
+
+    #[test]
+    fn predict_and_update_errors() {
+        let mut rng = SimpleRandom::new();
+        let header = GroupHeader {
+            use_global_tree: false,
+            wp_header: WeightedHeader {
+                all_default: true,
```

Comment:
This seems inconsistent.

Also, these are clearly invalid values, we should be specifying values could be read from a bitstream (so up to 5 bits for p..., and up to 4 bits for w...), as otherwise we are testing behaviour we don't care about.

Reply:
It matters, because it might not *stay* the same in the future when we do optimizations.
Also, Rust in debug mode panics on overflow, which these values would certainly cause at some point or another.

Reply:
It would help, yes.

Reply:
Done